### ISSUE TITLE
Gradient checkpointing for `grad_weight` in LFCE

### ISSUE THREAD

**Issue by cassanof** [linkedin/Liger-Kernel#533]
### 🚀 The feature, motivation and pitch

The LFCE kernel allocates a `grad_weight` tensor:
https://github.com/linkedin/Liger-Kernel/blob/a8fa3bb37850e89500261024ff47da0c626ab75f/src/liger_kernel/ops/fused_linear_cross_entropy.py#L47
This tensor then gets updated throughout the chunked loss calculation and finally used in the backward as a custom grad operation:
https://github.com/linkedin/Liger-Kernel/blob/a8fa3bb37850e89500261024ff47da0c626ab75f/src/liger_kernel/ops/fused_linear_cross_entropy.py#L127-L136

This has shape `[vocab_size, seq_len]`, which in some situations can become very large, which makes it impossible to do large pipeline parallel microbatches at long sequence lengths, as I have to keep this tensor in memory until the backward. It would be great to have gradient check-pointing here.

**Comment by cassanof** [comment]
I wrote a fully recomputed version of LFCE which resolves this issue. Would other people want it? I guess not many are training big models on long context.

**Comment by ByronHsu** [comment]
Hi @cassanof this sounds reasonable and seems that it can further reduce the memory. Have you measured the speed as it may introduce some regression? We greatly appreciate if you are willing to contribute the code. Some users do raise similar requests

**Comment by cassanof** [comment]
Hi! Recomputation definitely regresses throughput, it's only useful if you are doing pipeline parallelism at a high degree.
In my testing, it has regressed throughput by 15% with the same batch size. Though, due to the memory savings on the last rank, i can now 4x the batch size, which has given me a perf improvement! With pipeline parallelism, I can also balance the layers to earlier ranks, so that the regression in throughput is fully amortized. Lmk if this addition is wanted, would probably need to be a separate loss function altogether.

**Comment by ByronHsu** [comment]
Cool! Feel free to share us with a draft. I can find folks to polish and merge into liger. cc @austin362667 @Tcc0403
